### PR TITLE
docs:  New terms and updates to glossary

### LINKED ISSUE #21801: Adding additional terms to the Transformers glossary
### Feature request

Adding definitions to the [Transformers glossary](https://huggingface.co/docs/transformers/glossary) for each of the following terms:

- **encoder, decoder and encoder/decoder:** You already have definitions for autoencoding and autoregressive models but it's not immediately clear from the glossary that those are synonymous to encoder and decoder. Could point to relevant sections of the [Summary of the models](https://huggingface.co/docs/transformers/model_summary) article.
- **finetuned model:** There already is a glossary term for pretrained model, can link to the [Fine-tune a pretrained model](https://huggingface.co/docs/transformers/training) article
- **inference**  
- **pipeline:** Could link to the [Pipelines for inference](https://huggingface.co/docs/transformers/pipeline_tutorial) article
- **preprocessing:** Can link to [the preprocess document](https://huggingface.co/docs/transformers/preprocessing)
- **supervised and unsupervised learning** 

A few other definitions which might be worth defining (even if they don't show up as much in your documentation) is  **representation learning**, **semi-supervised learning**, **feature extraction**, **Large Language Models (LLM)** (mainly because it is such a popular term now) and **transfer learning**.

Finally, it might be worth putting acronyms beside glossary terms like **natural language processing/understanding** and **recurrent neural networks** (i.e. NLP/U and RNN) both for brevity and because they are so commonly used.

### Motivation

The glossary page has been quite helpful for me in understanding certain overloaded terms in deep learning, and I feel that adding these terms would be beneficial to others. It also could help link people to useful articles as many of the above terms have been explained already in one of your articles which helps with keeping things organized. 

### Your contribution

I would be happy to help with coming up with the definitions and submitting a PR with the added changes 🙂

### PR DESCRIPTION
**Pull request by MichaelRipa**
Updates to the glossary as proposed in #21801 .

All terms suggested in the issue were added, except for Encoder/Decoder. Acronyms were also added for terms in which it made sense (e.g. NLP)

Note that this PR replaces ***autoencoding models*** with ***encoder models*** (and provides a more detailed definition) as well as merging ***autoregressive models*** and ***causal language modeling*** with ***decoder models***.

This is my first draft and would likely benefit from additional edits/revision. Any suggestions in terms of updates or areas to expand further on are welcome 🙂 

# What does this PR do?

Fixes #21801 

## Before submitting
- [x ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@stevhliu

### CODE REVIEW

File: docs/source/en/glossary.mdx [hunk 1]
```diff
@@ -269,9 +283,13 @@ about their specific labels!
 The base models ([`BertModel`]) do not accept labels, as these are the base transformer models, simply outputting
 features.
 
+### large language models (LLM)
+
+A generic term which refers to transformer language models (e.g. GPT-3, BERT) that were trained on a large quantity of data.
```
Comment (stevhliu) [inline docs/source/en/glossary.mdx:288]: These might be more relevant examples of LLMs than BERT

```suggestion
A generic term that refers to transformer language models (GPT-3, BLOOM, OPT) that were trained on a large quantity of data.
```

File: docs/source/en/glossary.mdx [hunk 2]
```diff
@@ -282,21 +300,27 @@ A task that combines texts with another kind of inputs (for instance images).
...
+### Natural language understanding (NLU)
 
 All tasks related to understanding what is in a text (for instance classifying the
 whole text, individual words).
 
 ## P
 
+### pipeline
+
+A pipeline is an abstraction referring to a series of steps that are executed in a specific order to process and transform data to be used by a downstream machine learning model for training or inference. Some example stages found in a pipeline might be data preprocessing, feature extraction, normalization and model training.
```
Comment (stevhliu) [inline docs/source/en/glossary.mdx:320]: Pipelines in the context of 🤗 Transformers don't involve training, so it'd probably be good to remove that to avoid any confusion.

```suggestion
A pipeline in 🤗 Transformers is an abstraction referring to a series of steps that are executed in a specific order to preprocess and transform data and return a prediction from a model. Some example stages found in a pipeline might be data preprocessing, feature extraction, and normalization.
```

File: docs/source/en/glossary.mdx [hunk 3]
```diff
@@ -282,21 +300,27 @@ A task that combines texts with another kind of inputs (for instance images).
...
+
+For more details see [Pipelines for inference](https://huggingface.co/docs/transformers/pipeline_tutorial)
```
Comment (stevhliu) [inline docs/source/en/glossary.mdx:322]:
```suggestion
For more details, see [Pipelines for inference](https://huggingface.co/docs/transformers/pipeline_tutorial).
```

File: docs/source/en/glossary.mdx [hunk 4]
```diff
@@ -317,6 +341,9 @@ absolute positional embeddings.
 Absolute positional embeddings are selected in the range `[0, config.max_position_embeddings - 1]`. Some models use
 other types of positional embeddings, such as sinusoidal position embeddings or relative position embeddings.
 
+### preprocessing
+
+The task of cleaning and organizing raw data into a format that can be consumed easily by machine learning models. To gain a better idea of what preprocessing is, we recommend checking out the [Preprocess](https://huggingface.co/docs/transformers/preprocessing) article, which goes over a large range of examples.
```
Comment (stevhliu) [inline docs/source/en/glossary.mdx:346]:
```suggestion
The task of preparing raw data into a format that can be easily consumed by machine learning models. For example, text is typically preprocessed by tokenization. To gain a better idea of what preprocessing looks like for other input types, check out the [Preprocess](https://huggingface.co/docs/transformers/preprocessing) tutorial.
```

File: docs/source/en/glossary.mdx [hunk 5]
```diff
@@ -329,10 +356,14 @@ modeling](#masked-language-modeling)).
...
 ## R
 
-### recurrent neural network
+### recurrent neural network (RNN)
 
 A type of model that uses a loop over a layer to process texts.
 
+### representation learning
+
+A subfield of machine learning which focuses on learning meaningful representations of raw data. Some examples of representation learning techniques include the use of word embeddings, autoencoders, and Generative Adversarial Networks (GANs).
```
Comment (stevhliu) [inline docs/source/en/glossary.mdx:365]:
```suggestion
A subfield of machine learning which focuses on learning meaningful representations of raw data. Some examples of representation learning techniques include word embeddings, autoencoders, and Generative Adversarial Networks (GANs).
```

File: docs/source/en/glossary.mdx [hunk 6]
```diff
@@ -343,6 +374,12 @@ A measurement in hertz of the number of samples (the audio signal) taken per sec
 
 Each element of the input finds out which other elements of the input they should attend to.
 
+### semi-supervised learning
+
+A broad category of machine learning training techniques which look to leverage a small amount of labeled data with a larger quantity of unlabeled data to improve the accuracy of a model, differentiating it from [supervised learning](#supervised-learning) and [unsupervised learning](#unsupervised-learning).
```
Comment (stevhliu) [inline docs/source/en/glossary.mdx:379]:
```suggestion
A broad category of machine learning training techniques that leverages a small amount of labeled data with a larger quantity of unlabeled data to improve the accuracy of a model, unlike [supervised learning](#supervised-learning) and [unsupervised learning](#unsupervised-learning).
```

File: docs/source/en/glossary.mdx [hunk 7]
```diff
@@ -352,6 +389,10 @@ Models that generate a new sequence from an input, like translation models, or s
 
 In [convolution](#convolution) or [pooling](#pooling), the stride refers to the distance the kernel is moved over a matrix. A stride of 1 means the kernel is moved one pixel over at a time, and a stride of 2 means the kernel is moved two pixels over at a time.
 
+### supervised learning
+
+A form of model training which makes direct use of labeled data to correct and instruct model performance. Data is fed into the model being trained, and its predictions are then compared to the known labels. The model then has its weights updated based on how incorrect its predictions were, and the process is repeated to try and optimize model performance.
```
Comment (stevhliu) [inline docs/source/en/glossary.mdx:394]:
```suggestion
A form of model training that directly uses labeled data to correct and instruct model performance. Data is fed into the model being trained, and its predictions are compared to the known labels. The model updates its weights based on how incorrect its predictions were, and the process is repeated to optimize model performance.
```

File: docs/source/en/glossary.mdx [hunk 8]
```diff
@@ -410,6 +451,16 @@ sequence, corresponding to the "question", has all its tokens represented by a `
 
 Some models, like [`XLNetModel`] use an additional token represented by a `2`.
 
+### transfer learning
+
+A technique which involves taking a model pretrained for a specific task and using it for something completely different. Instead of training a model from scratch, we can leverage knowledge obtained from an existing model as a starting point, which can speed up the learning process and reduce the amount of training data needed.
```
Comment (stevhliu) [inline docs/source/en/glossary.mdx:456]:
```suggestion
A technique that involves taking a pretrained model and adapting it to a dataset specific to your task. Instead of training a model from scratch, you can leverage knowledge obtained from an existing model as a starting point. This speeds up the learning process and reduces the amount of training data needed.
```

**Review by stevhliu — COMMENTED** [review]
Thanks for adding these new terms! I added some suggestions for clarity 🙂

To get the CircleCI tests to pass, you may need to refresh your permissions as shown [here](https://support.circleci.com/hc/en-us/articles/360048210711-How-to-Refresh-User-Permissions-).

File: docs/source/en/glossary.mdx [hunk 9]
```diff
@@ -160,6 +170,10 @@ The model head refers to the last layer of a neural network that accepts the raw
 
 Vision-based Transformers models split an image into smaller patches which are linearly embedded, and then passed as a sequence to the model. You can find the `patch_size` - or resolution - of the model in it's configuration.
 
+### inference
+
+Inference is the process of evaluating a model on new data after the training process is complete.
```
[the remark this answers is not in the document]
Comment (MichaelRipa) [reply docs/source/en/glossary.mdx]: That's a good idea, I'll add that in

File: docs/source/en/glossary.mdx (hunk 2 above)
Comment (MichaelRipa) [reply docs/source/en/glossary.mdx:322]: Oh yeah good catch!

**Comment by MichaelRipa** [comment]
Thanks for the detailed edits, they are helpful! 🙂 I will refresh permissions and also add a link to the pipeline for inference doc tomorrow

**Comment by MichaelRipa** [comment]
Added my update, and I followed the instructions to refresh permissions. I don't seem to have permissions to manually restart the CircleCI Pipeline, so not sure whether the steps I took were reflected in my last commit (I did my updates before refreshing permissions).

**Comment by sgugger** [comment]
You can push an empty commit with `git commit -m "Trigger CI" --allow-empty" and then push to your branch.

File: docs/source/en/glossary.mdx [hunk 10]
```diff
@@ -282,21 +302,27 @@ A task that combines texts with another kind of inputs (for instance images).
 
 ## N
 
-### Natural language generation
+### Natural language generation (NLG)
 
 All tasks related to generating text (for instance talk with transformers, translation).
```
Comment (stevhliu) [inline docs/source/en/glossary.mdx:307]: Did you mean Write With Transformers?

```suggestion
All tasks related to generating text (for instance, [Write With Transformers](https://transformer.huggingface.co/), translation).
```

**Review by stevhliu — APPROVED** [review]
Thanks so much for your work, I left one more comment, but I think this should be good to go! Pinging @sgugger for a final look :)

File: docs/source/en/glossary.mdx [hunk 11]
```diff
@@ -73,14 +73,6 @@ by the tokenizer under the key "attention_mask":
 [[1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]
 ```
 
-### autoencoding models 
-
-see [masked language modeling](#masked-language-modeling)
-
-### autoregressive models
-
-see [causal language modeling](#causal-language-modeling)
```
Comment (sgugger) [inline docs/source/en/glossary.mdx:82]: I would keep those as these are sometimes used (a quick search in the repo shows quite a few instances).

File: docs/source/en/glossary.mdx [hunk 12]
```diff
@@ -93,11 +85,6 @@ The backbone is the network (embeddings and layers) that outputs the raw hidden
 
 Color images are made up of some combination of values in three channels - red, green, and blue (RGB) - and grayscale images only have one channel. In 🤗 Transformers, the channel can be the first or last dimension of an image's tensor: [`n_channels`, `height`, `width`] or [`height`, `width`, `n_channels`].
 
-### causal language modeling
-
-A pretraining task where the model reads the texts in order and has to predict the next word. It's usually done by
-reading the whole sentence but using a mask inside the model to hide the future tokens at a certain timestep.
```
Comment (sgugger) [inline docs/source/en/glossary.mdx:99]: This definitely need to be kept, we use it all the time in the course and the examples. The class is `AutoModelForCausalLM` etc.

File: docs/source/en/glossary.mdx [hunk 13]
```diff
@@ -269,9 +285,13 @@ about their specific labels!
 The base models ([`BertModel`]) do not accept labels, as these are the base transformer models, simply outputting
 features.
 
+### large language models (LLM)
+
+A generic term that refers to transformer language models (GPT-3, BLOOM, OPT) that were trained on a large quantity of data.
```
Comment (sgugger) [inline docs/source/en/glossary.mdx:290]: Also linked to their size (in number of parameters) I believe.

File: docs/source/en/glossary.mdx [hunk 14]
```diff
@@ -343,6 +376,12 @@ A measurement in hertz of the number of samples (the audio signal) taken per sec
 
 Each element of the input finds out which other elements of the input they should attend to.
 
+### semi-supervised learning
+
+A broad category of machine learning training techniques that leverages a small amount of labeled data with a larger quantity of unlabeled data to improve the accuracy of a model, unlike [supervised learning](#supervised-learning) and [unsupervised learning](#unsupervised-learning).
+
+An example of an semi-supervised learning approach is "self-training", in which a model is trained on labeled data, and then used to make predictions on the unlabeled data. The portion of the unlabeled data that the model predicts with the most confidence gets added to the labeled dataset and used to retrain the model.
```
Comment (sgugger) [inline docs/source/en/glossary.mdx:383]:
```suggestion
An example of a semi-supervised learning approach is "self-training", in which a model is trained on labeled data, and then used to make predictions on the unlabeled data. The portion of the unlabeled data that the model predicts with the most confidence gets added to the labeled dataset and used to retrain the model.
```

**Review by sgugger — COMMENTED** [review]
Thanks for your PR. I don't think we should remove entries from the glossary. Linking to other entries is better. Could you also add an entry for "self-supervised learning" since most pretraining of Transformer models use that technique?

**Comment by MichaelRipa** [comment]
Thanks for the review! I'll start going over the changes / implementing your suggestions over the weekend

**Comment by MichaelRipa** [comment]
> Thanks for your PR. I don't think we should remove entries from the glossary. Linking to other entries is better. Could you also add an entry for "self-supervised learning" since most pretraining of Transformer models use that technique?

I will add back in the old entries and add links between them, and yeah that's a great idea!

File: docs/source/en/glossary.mdx (hunk 10 above)
Comment (MichaelRipa) [reply docs/source/en/glossary.mdx:307]: I didn't write this entry but good catch!

**Review by sgugger — APPROVED** [review]
Thanks for iterating! I believe there is one last suggestion from Steven to treat and we should be good to merge!

**Comment by MichaelRipa** [comment]
Awesome! I committed the suggestion, looks to have merged successfully🙂 Thanks again for the help with the edits/reviews! (also feel free to ping me if you see any further edits/additions to be done)
